### PR TITLE
REF: share code for scalar validation in datetimelike array methods

### PR DESCRIPTION
**Pull request by jbrockmendel**
The user-facing effect here is in the exception messages we show to users, where we may want to be more verbose than this.  Thoughts @gfyoung?

### CODE REVIEW

**Comment by gfyoung** [comment]
> The user-facing effect here is in the exception messages we show to users, where we may want to be more verbose than this. Thoughts @gfyoung?

I agree that we would want to improve the user-friendliness of these error messages.  From that perspective, I prefer the error messages that were there before.

File: pandas/core/arrays/datetimelike.py [hunk 1]
```diff
@@ -798,72 +818,35 @@ def _validate_listlike(
...
-        elif isinstance(value, self._recognized_scalars):
-            value = self._scalar_type(value)
-        elif is_valid_nat_for_dtype(value, self.dtype):
-            value = NaT
         else:
-            msg = (
-                f"'value' should be a '{self._scalar_type.__name__}', 'NaT', "
-                f"or array of those. Got '{type(value).__name__}' instead."
-            )
-            raise TypeError(msg)
```
Comment (gfyoung) [inline pandas/core/arrays/datetimelike.py:838]: This message for example was a nice user-friendly one.

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -760,6 +759,26 @@ def _validate_shift_value(self, fill_value):
 
         return self._unbox(fill_value)
 
+    def _validate_scalar(self, value, msg: str, cast_str: bool = False):
+        if cast_str and isinstance(value, str):
```
Comment (jreback) [inline pandas/core/arrays/datetimelike.py:763]: can you add a doc-string here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/datetimelike.py (hunk 1 above)
Comment (jreback) [reply pandas/core/arrays/datetimelike.py:838]: ok @gfyoung ?

File: pandas/core/arrays/datetimelike.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/datetimelike.py:775]: updated

**Review by jreback — APPROVED** [review]

**Pull request merged**